File: back/src/main/java/com/back/domain/review/validation/ValidTagsValidator.java
package com.back.domain.review.validation;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

public class ValidTagsValidator implements ConstraintValidator<ValidTags, List<String>> {
    private List<String> value;

    @Override
    public boolean isValid(List<String> value, ConstraintValidatorContext context) {
        if (value == null) return true;

        this.value = value;

        return isMaxSize5() && isAllTagValid() && isAllTagUnique();
    }

    private boolean isMaxSize5() {
        return value.size() <= 5;
    }

    private boolean isAllTagValid() {
        for (int i = 0; i < value.size(); i++) {
            String tag = value.get(i);
            if (1 <= tag.length() && tag.length() <= 30) continue;
            return false;
        }

        return true;
    }

    private boolean isAllTagUnique() {
        Map<String, Integer> tagCounts = new HashMap<>();

        for (int i = 0; i < value.size(); i++) {
            String tag = value.get(i);
            tagCounts.put(tag, Optional.ofNullable(tagCounts.get(tag)).orElse(0) + 1);
        }

        for (int count : tagCounts.values()) {
            if (count <= 1) continue;
            return false;
        }

        return true;
    }
}

File: back/src/main/java/com/back/domain/review/validation/ValidTagsValidator.kt
package com.back.domain.review.validation

import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext

class ValidTagsValidator : ConstraintValidator<ValidTags, List<String>> {
    private var value: List<String>? = null

    override fun isValid(value: List<String>?, context: ConstraintValidatorContext?): Boolean {
        if (value == null) return true

        this.value = value

        return this.isMaxSize5 && this.isAllTagValid && this.isAllTagUnique
    }

    private val isMaxSize5: Boolean
        get() = value!!.size <= 5

    private val isAllTagValid: Boolean
        get() {
            for (i in value!!.indices) {
                val tag = value!![i]
                if (tag.length in 1..30) continue
                return false
            }

            return true
        }

    private val isAllTagUnique: Boolean
        get() {
            val tagCounts: MutableMap<String, Int> =
                mutableMapOf()

            for (i in value!!.indices) {
                val tag = value!![i]
                tagCounts[tag] = (tagCounts[tag] ?: 0) + 1
            }

            for (count in tagCounts.values) {
                if (count <= 1) continue
                return false
            }

            return true
        }
}